### ISSUE TITLE
Emit an event when database has finished syncing with connected peers

### ISSUE THREAD

**Issue by tyleryasaka** [orbitdb/orbitdb#509]
https://github.com/orbitdb/orbit-db/issues/474 describes a situation in which this event would be useful. Existing events do not communicate this information.

The `load` event is emitted simply once a database has finished loading from memory. This event may be emitted before the database has synced with connected peers.

The `replicated` event is in fact emitted after a database has synced; however, it does not always fire (and there is no straightforward way to determine whether it will fire). For example, if a connected peer does not have the database open, or it does not have any heads for the database, then no replication will occur and the `replicated` event will not fire. Thus, it's uncertain whether the `replicated` event will ever fire, so one cannot reliably wait for this event to fire in order to know that a database has finished syncing with peers.

I have implemented a solution for this problem; however, this solution is not sufficient in its current form. I'm linking to it here as an example of what a solution might look like. The solution spans 2 repos:

[orbit-db changes](https://github.com/orbitdb/orbit-db/compare/master...projectaspen:synced-event)
[orbit-db-store changes](https://github.com/orbitdb/orbit-db-store/compare/master...projectaspen:wait-process-queue)

Further, here is an example of code that uses this new `synced` event to know when a database has finished syncing (usage is still a bit complex, as the event fires only once per peer/database combination, and requires additional pubsub message sending to request that a connected peer open a specific databse): https://github.com/projectaspen/aspen/blob/master/packages/orbit-connect-client/src/orbit-connect-client.js#L104

**Comment by haadcode** [comment]
@tyleryasaka we discussed this yesterday in the community call. See [the notes here](https://github.com/orbitdb/welcome/blob/master/meeting-notes/2018/2018.11.27.md) (under "semantics of synced"). The notes should capture the (design/ux) challenge here, but if they don't, let us know and we can continue the discussion here.

**Comment by tyleryasaka** [comment]
I've created a PR that implements the solution I am referring to. It makes minimal code changes, and does not touch the enqueueing system (unlike my original solution). It requires an accompanying change to orbit-db-store.

https://github.com/orbitdb/orbit-db/pull/514
https://github.com/orbitdb/orbit-db-store/pull/38

**Comment by tyleryasaka** [comment]
Upon further investigation, I see that my initial approach was naive. The store `sync` method waits for heads to be replicated, but not necessarily the rest of the graph. I updated my PR to orbit-db-store to use the `replicated.progress` event to ensure that all of the `next`s have been followed and the db has been fully replicated. This seems to solve the problem, though it's not a perfect solution.

The downside to this is that if other things are being added to the queue rapidly, then this may hang indefinitely. It waits for *all* queue items to complete, not just the ones derived from the heads in the current request. That said, currently as this is used in orbit-db, this would not block anything other than the new event I have proposed adding. So it should be acceptable, maybe?

Ideally, what should really happen is that we should have a way of knowing when the items derived from the given heads have been replicated. Currently, however, the queue is a black hole and there's no way to track the graph that branches out from the heads. The queue would likely need to be refactored for this approach.

**Comment by haadcode** [comment]
@tyleryasaka reached out on Gitter to discuss this further. I like the general direction of the PRs, but we should think about covering all the different use cases and make sure we get the semantics right (ie. "what/when is ready?").

Thanks for pushing this! Fixing it would help many 👍

**Comment by holmesworcester** [comment]
Something like this would be helpful to us too! 

Specifically, we'd like to know when we've finished "syncing" with all peers after first connecting. I'm not completely clear on what the knowable things here are, but as was said above, it's fine that there's no finality, but "I know what the peers I'm connected to know as of, say, 30 seconds ago" is very valuable, e.g. for users who have been offline and are coming back online, or for users who might suspect they have some issue with their internet connection.

**Comment by haydenyoung** [comment]
Due to the archiving of many of the legacy modules and with the release of version 1, the attached PR will, unfortunately, be stale.

I'm going to close this issue as part of housekeeping but feel free to move the discussion to a new issue including solutions in the context of v1.
